### ISSUE TITLE
ReferenceError: ..._node_modules_babel_runtime_corejs2_core_js_promise__WEBPACK_IMPORTED_MODULE_0___default is not defined

### ISSUE THREAD

**Issue by SaphuA** [michaelolof/vuex-class-component#4]
Thanks to your feedback and the recent @getter bug fix I am able to compile my code. However I am now running into a runtime error that I was easily able to reproduce in my test repo.

https://github.com/SaphuA/vuex-class-component-test/blob/6f545a82c3cb1b8132505bee27d5d324b08c2386/test/src/App.vue#L27

The error is thrown as soon as I try to call an (empty) action method in my store.

https://github.com/SaphuA/vuex-class-component-test/blob/6f545a82c3cb1b8132505bee27d5d324b08c2386/test/src/store/athena/index.ts#L13

I'm currently researching if the issue is caused by webpack in combination with the circular references in your lib, but it could be something completely different...

**Comment by michaelolof** [comment]
Read the docs on Actions.

Actions are tricky, because it uses context. This is why action comes in two modes, `mutate` or `raw`.
The default is mutate and based on your build process it is not guaranteed to work all the time.

This is acutally a babel problem. The mutate mode attempts to resolve all your `this` calls to their respective `context.state`, `context.commit`, `context.dispatch` etc The problem is that babel introduces its own transpiled constant into your code which loses scope after the actions `this` calls have been resolved (I'm still trying to fix this or come up with a better alternative)

But this is why the `raw` mode exists.
Generally as a rule of thumb is: If you use the `mutate` mode and get errors, switch to the `raw` mode. That is guaranteed to work.

So with the `raw` mode it should look like this.
```ts
@action({ mode: "raw" })
public readUzovis() {
  return Promise.resolve();
}
```
This would work.

And if you want to access state or getters, you get back the context object.

```ts
import { getRawActionContext } from "vuex-class-component";

...

@action({ mode: "raw" })
public readUzovis() {
  const context = getRawActionContext( this );

  context.state.foo // to access your state like you would in regular vuex.
  context.getters.foo // to access your getters.
  context.commit('setFoo', 'bar'); // to access your mutations.
  context.dispatch('asyncStuff').then( val => console.log( val) ); // For your actions.

  return Promise.resolve();
}
```
In raw actions you don't use `this`

**Comment by SaphuA** [comment]
So then this does mean that there is no way to get a 'mutate' action that only does a couple of mutations? Since an action must always return a promise. This is why I ended up returning Promise.resolve()...

**Comment by michaelolof** [comment]
Honestly
Its really all about your build process.

I personally don't use babel with my ts-loader. I configure my ts-loader to talk directly with webpack. Babel is redundant here IMHO.

For now, if you remove babel from your build process, it would work. (In the future you may not have to)

> I'm not a huge fan of the getRawActionContext hack.

Think of it as more of a fallback rather than an hack.

**Comment by SaphuA** [comment]
Yeah getting rid of babel is on my todo list, but this is existing software and I don't want to introduce too many changes at once.

I feel like the mutate-actions are the way to go with this library since the store itself should be strongly typed as well. Are you planning on fixing this is some day or is it an unfixable dead end?

**Comment by SaphuA** [comment]
@michaelolof
I spent more time than I would like to admit moving from babel/vue-cli to plain old webpack. Unfortunately the mutate actions are still causing the same problems. Having to use raw actions, especially for the simpler ones, is kind of a downer. I guess it's better than not having typings in the view, but not where I was hoping to get.

I'm having problems understanding the challenges you have with raw vs mutate actions. Is there any way I could help?

**Comment by michaelolof** [comment]
Funny, I actually have a fix for this.

Just working on optimizing it without breaking the api.

**Comment by michaelolof** [comment]
I'll be pushing two new updates.

`v 1.4.0` which fixes the existing problem with the mutate mode and
`v 2.0.0` which deprecates both `mutate` and `raw` mode altogether

**Comment by michaelolof** [comment]
Fix has been added at v 1.4.0.

You can update and try it now.

**Comment by SaphuA** [comment]
Awesome it seems to be working perfectly! That makes this awesome lib a mandatory install on all my future Vue projects.
